src/dataflow/src/storage/render/debezium.rs

```diff
@@ -118,6 +121,304 @@ where
+    envelope: &DebeziumEnvelope,
+    input: &Stream<G, DecodeResult>,
+    tx_ok: Collection<G, Row, Diff>,
+    debug_name: String,
+) -> (
+    Stream<G, (Row, Timestamp, Diff)>,
+    Stream<G, (mz_dataflow_types::DataflowError, Timestamp, Diff)>,
+)
+where
+    G: ScopeParent<Timestamp = Timestamp>,
+{
+    let (before_idx, after_idx) = (envelope.before_idx, envelope.after_idx);
+    let hashed_id = data_id.hashed();
+
+    let (transaction_idx, tx_metadata_description) = match envelope {
+        DebeziumEnvelope {
+            mode:
+                DebeziumMode::Full(DebeziumDedupProjection {
+                    transaction_idx,
+                    tx_metadata,
+                    ..
+                }),
+            ..
+        }
+        | DebeziumEnvelope {
+            mode:
+                DebeziumMode::Ordered(DebeziumDedupProjection {
+                    transaction_idx,
+                    tx_metadata,
+                    ..
+                }),
+            ..
+        }
+        | DebeziumEnvelope {
+            mode:
+                DebeziumMode::FullInRange {
+                    projection:
+                        DebeziumDedupProjection {
+                            transaction_idx,
+                            tx_metadata,
+                            ..
+                        },
+                    ..
+                },
+            ..
+        } => (
+            *transaction_idx,
+            tx_metadata
+                .clone()
+                .expect("render_tx should only be called when there's a transaction"),
+        ),
+        DebeziumEnvelope {
+            mode: DebeziumMode::None,
+            ..
+        } => panic!("render_tx should be called with a dedup mode"),
+    };
+
+    // XXX(chae): Should we try to avoid moving everything to one worker?
+    tx_ok
+        .inner
+        .binary_frontier(
+            &input,
+            Exchange::new(move |_| hashed_id),
+            Exchange::new(move |_| hashed_id),
+            "envelope-debezium-tx",
+            {
+                let envelope = envelope.clone();
+                let mut tx_data = vec![];
+                let mut data = vec![];
+                let mut data_buffer = VecDeque::new();
+                let mut dedup_state = HashMap::new();
+
+                // Keep mapping of `transaction_id`s to the timestamp at which that transaction END record was read from
+                // the transaction metadata stream.  That stored timestamp will be the timestamp for all data records
+                // with the corresponding `transaction_id`.
+                let mut tx_mapping = HashMap::new();
+                // Hold onto a capability for each `transaction_id`.  This represents the time at which we'll emit
+                // matched data rows.  This will be dropped when we've matched the indicated number of events.
+                let mut tx_cap_map = HashMap::new();
+                let mut tx_event_count = HashMap::new();
```

Comment:
We should probably keep a single map that maps a transaction id to a `(time, cap, event_count)` tuple. Below in the logic these three maps are populated at the same time and the existence of a value in one is used as evidence for the existence of a value in the other ones (the places where we panic on Vacant entries). If this was a single map where all the information is inserted and extracted in one place there would be no possibility for misuse.